File: backend/src/services/user.service.ts
import { PrismaClient } from "../generated/prisma/index.js";
import type { User } from "../generated/prisma/index.js";
import logger from "../utils/logger.js";
import { createUserStorageFolders } from "../utils/user-storage.js";

const prisma = new PrismaClient();

interface CreateUserData {
  supabaseUid: string;
  email: string;
  displayName?: string;
  photoUrl?: string;
}

interface UpdateProfileData {
  displayName?: string;
  bio?: string;
  interests?: string[];
  socialLinks?: Record<string, string>;
  age?: number;
  gender?: string;
  photoUrl?: string;
  isOnboardingCompleted?: boolean;
}

export const createUser = async (data: CreateUserData): Promise<User> => {
  try {
    logger.info(`Creating user with UID: ${data.supabaseUid}`);
    const existingUser = await prisma.user.findUnique({
      where: { email: data.email },
    });
    if (existingUser) {
      logger.info(`User already exists with email: ${data.email}`);
      // Создаём папки для существующего пользователя (если их ещё нет)
      createUserStorageFolders(existingUser.id);

      // Если пользователь существует, но у него нет supabaseUid (например, старая запись), обновим его
      if (!existingUser.supabaseUid) {
        logger.info(
          `Updating existing user ${existingUser.id} with supabaseUid: ${data.supabaseUid}`,
        );
        return await prisma.user.update({
          where: { id: existingUser.id },
          data: { supabaseUid: data.supabaseUid },
        });
      }
      // Если supabaseUid уже есть и совпадает - вернем пользователя
      if (existingUser.supabaseUid === data.supabaseUid) {
        return existingUser;
      }
      // Если supabaseUid отличается - это конфликт
      throw new Error("User with this email already exists");
    }
    const newUser = await prisma.user.create({
      data: {
        supabaseUid: data.supabaseUid,
        email: data.email,
        displayName: data.displayName || "",
        photoUrl: data.photoUrl || "",
      },
    });
    logger.info(`User created with UID: ${data.supabaseUid}`);

    // Создаём папки для хранения файлов пользователя
    createUserStorageFolders(newUser.id);

    return newUser;
  } catch (error) {
    logger.error(`Error creating user with UID: ${data.supabaseUid}`, error);
    throw error;
  }
};

export const getUserBySupabaseUid = async (
  supabaseUid: string,
): Promise<User | null> => {
  logger.info(`Fetching user with UID: ${supabaseUid}`);
  try {
    const user = await prisma.user.findUnique({ where: { supabaseUid } });
    if (user) {
      logger.info(`User found with UID: ${supabaseUid}`);
    } else {
      logger.warn(`No user found with UID: ${supabaseUid}`);
    }
    return user;
  } catch (error) {
    logger.error(`Error fetching user with UID: ${supabaseUid}`, error);
    throw error;
  }
};

export const getUserById = async (userId: string): Promise<User | null> => {
  logger.info(`Fetching user with ID: ${userId}`);
  try {
    const user = await prisma.user.findUnique({ where: { id: userId } });
    if (user) {
      logger.info(`User found with ID: ${userId}`);
    } else {
      logger.warn(`No user found with ID: ${userId}`);
    }
    return user;
  } catch (error) {
    logger.error(`Error fetching user with ID: ${userId}`, error);
    throw error;
  }
};

export const updateUserProfile = async (
  userId: string,
  data: UpdateProfileData,
): Promise<User> => {
  logger.info(`Updating profile for user ID: ${userId}`);
  try {
    const user = await prisma.user.update({
      where: { id: userId },
      data: { ...data },
    });
    return user;
  } catch (error) {
    logger.error(`Error updating profile for user ID: ${userId}`, error);
    throw error;
  }
};

export const updateUserLocation = async (
  userId: string,
  latitude: number,
  longitude: number,
): Promise<void> => {
  logger.info(`Updating location for user ID: ${userId}`);
  try {
    await prisma.user.update({
      where: { id: userId },
      data: {
        lastLatitude: latitude,
        lastLongitude: longitude,
        lastLocationUpdate: new Date(),
      },
    });
    logger.info(`Location updated for user ID: ${userId}`);
  } catch (error) {
    logger.error(`Error updating location for user ID: ${userId}`, error);
    throw error;
  }
};

export const getMatchesForUser = async (
  userId: string,
  latitude?: number,
  longitude?: number,
  radiusKm: number = 50,
  limit: number = 20,
): Promise<User[]> => {
  logger.info(`Fetching matches for user ID: ${userId}`);
  try {
    // Получаем текущего пользователя для получения его фильтров
    const currentUser = await prisma.user.findUnique({
      where: { id: userId },
    });

    if (!currentUser) {
      logger.warn(`User not found: ${userId}`);
      return [];
    }

    // Используем переданные координаты или последние известные координаты пользователя
    const userLat = latitude ?? currentUser.lastLatitude;
    const userLon = longitude ?? currentUser.lastLongitude;

    // Если координаты не доступны, возвращаем пустой список
    if (!userLat || !userLon) {
      logger.warn(`No location available for user ${userId}`);
      return [];
    }

    // Вычисляем границы поиска (приближённо)
    const radiusMeters = radiusKm * 1000;
    const earthRadiusKm = 6371;
    const latChange = (radiusKm / earthRadiusKm) * (180 / Math.PI);
    const lonChange =
      (radiusKm / (earthRadiusKm * Math.cos((userLat * Math.PI) / 180))) *
      (180 / Math.PI);

    const minLat = userLat - latChange;
    const maxLat = userLat + latChange;
    const minLon = userLon - lonChange;
    const maxLon = userLon + lonChange;

    logger.info(
      `Search bounds: lat [${minLat}, ${maxLat}], lon [${minLon}, ${maxLon}]`,
    );

    // Получаем пользователей в пределах географической области
    let query: any = {
      where: {
        AND: [
          // Исключаем самого пользователя
          {
            id: {
              not: userId,
            },
          },
          // Исключаем пользователей, которые не завершили онбординг
          {
            isOnboardingCompleted: true,
          },
          // Исключаем невидимые профили
          {
            isProfileVisible: true,
          },
          // Геолокационный фильтр
          {
            lastLatitude: {
              gte: minLat,
              lte: maxLat,
            },
          },
          {
            lastLongitude: {
              gte: minLon,
              lte: maxLon,
            },
          },
        ],
      },
      take: limit,
    };

    // Если установлены возрастные ограничения текущего пользователя, применяем фильтр
    if (currentUser.minAge || currentUser.maxAge) {
      const ageFilter: any = {};
      if (currentUser.minAge) {
        ageFilter.gte = currentUser.minAge;
      }
      if (currentUser.maxAge) {
        ageFilter.lte = currentUser.maxAge;
      }

      query.where.AND.push({
        age: ageFilter,
      });
    }

    const matches = await prisma.user.findMany(query);

    logger.info(`Found ${matches.length} potential matches for user ${userId}`);
    return matches;
  } catch (error) {
    logger.error(`Error fetching matches for user ${userId}:`, error);
    throw error;
  }
};

File: legacy/go-services/users-service/internal/service/user_service.go
package service

import (
	"context"
	"errors"

	"go.uber.org/zap"

	"github.com/AnathanWang/andexevents/services/users-service/internal/model"
	"github.com/AnathanWang/andexevents/services/users-service/internal/repository"
)

// Ошибки сервиса
var (
	ErrUserNotFound       = errors.New("user not found")
	ErrUserAlreadyExists  = errors.New("user with this email already exists")
	ErrInvalidCoordinates = errors.New("invalid coordinates")
	ErrNoLocation         = errors.New("no location available")
)

// UserService интерфейс сервиса пользователей
type UserService interface {
	CreateUser(ctx context.Context, supabaseUID, email string, displayName, photoURL *string) (*model.User, error)
	GetCurrentUser(ctx context.Context, userID string) (*model.User, error)
	GetUserBySupabaseUID(ctx context.Context, supabaseUID string) (*model.User, error)
	UpdateProfile(ctx context.Context, userID string, req *model.UpdateUserRequest) (*model.User, error)
	UpdateLocation(ctx context.Context, userID string, latitude, longitude float64) error
	GetMatches(ctx context.Context, userID string, latitude, longitude *float64, radiusKm float64, limit int) ([]model.User, error)
}

type userService struct {
	repo   repository.UserRepository
	logger *zap.Logger
}

// NewUserService создаёт новый сервис пользователей
func NewUserService(repo repository.UserRepository, logger *zap.Logger) UserService {
	return &userService{
		repo:   repo,
		logger: logger,
	}
}

// CreateUser создаёт нового пользователя
func (s *userService) CreateUser(ctx context.Context, supabaseUID, email string, displayName, photoURL *string) (*model.User, error) {
	s.logger.Info("Creating user",
		zap.String("supabaseUID", supabaseUID),
		zap.String("email", email),
	)

	// Проверяем существует ли пользователь с таким email
	existingUser, err := s.repo.GetByEmail(ctx, email)
	if err != nil {
		s.logger.Error("Error checking existing user", zap.Error(err))
		return nil, err
	}

	if existingUser != nil {
		s.logger.Info("User already exists with email", zap.String("email", email))

		// Если supabaseUid пустой - обновляем
		if existingUser.SupabaseUID == "" {
			s.logger.Info("Updating existing user with supabaseUid",
				zap.String("userID", existingUser.ID),
				zap.String("supabaseUID", supabaseUID),
			)
			return s.repo.UpdateSupabaseUID(ctx, existingUser.ID, supabaseUID)
		}

		// Если supabaseUid совпадает - возвращаем пользователя
		if existingUser.SupabaseUID == supabaseUID {
			return existingUser, nil
		}

		// Если supabaseUid отличается - это конфликт
		return nil, ErrUserAlreadyExists
	}

	// Создаём нового пользователя
	user, err := s.repo.Create(ctx, supabaseUID, email, displayName, photoURL)
	if err != nil {
		s.logger.Error("Error creating user", zap.Error(err))
		return nil, err
	}

	s.logger.Info("User created successfully", zap.String("userID", user.ID))
	return user, nil
}

// GetCurrentUser получает текущего пользователя по ID из БД
func (s *userService) GetCurrentUser(ctx context.Context, userID string) (*model.User, error) {
	s.logger.Info("Getting user", zap.String("userID", userID))

	user, err := s.repo.GetByID(ctx, userID)
	if err != nil {
		s.logger.Error("Error getting user", zap.Error(err))
		return nil, err
	}

	if user == nil {
		s.logger.Warn("User not found", zap.String("userID", userID))
		return nil, ErrUserNotFound
	}

	return user, nil
}

// GetUserBySupabaseUID получает пользователя по Supabase UID
func (s *userService) GetUserBySupabaseUID(ctx context.Context, supabaseUID string) (*model.User, error) {
	s.logger.Info("Getting user by supabaseUID", zap.String("supabaseUID", supabaseUID))

	user, err := s.repo.GetBySupabaseUID(ctx, supabaseUID)
	if err != nil {
		s.logger.Error("Error getting user by supabaseUID", zap.Error(err))
		return nil, err
	}

	if user == nil {
		s.logger.Warn("User not found", zap.String("supabaseUID", supabaseUID))
		return nil, ErrUserNotFound
	}

	return user, nil
}

// UpdateProfile обновляет профиль пользователя
func (s *userService) UpdateProfile(ctx context.Context, userID string, req *model.UpdateUserRequest) (*model.User, error) {
	s.logger.Info("Updating user profile", zap.String("userID", userID))

	// Проверяем существование пользователя
	existing, err := s.repo.GetByID(ctx, userID)
	if err != nil {
		s.logger.Error("Error checking user existence", zap.Error(err))
		return nil, err
	}
	if existing == nil {
		return nil, ErrUserNotFound
	}

	user, err := s.repo.Update(ctx, userID, req)
	if err != nil {
		s.logger.Error("Error updating user profile", zap.Error(err))
		return nil, err
	}

	s.logger.Info("User profile updated successfully", zap.String("userID", userID))
	return user, nil
}

// UpdateLocation обновляет геолокацию пользователя
func (s *userService) UpdateLocation(ctx context.Context, userID string, latitude, longitude float64) error {
	s.logger.Info("Updating user location",
		zap.String("userID", userID),
		zap.Float64("latitude", latitude),
		zap.Float64("longitude", longitude),
	)

	// Валидация координат
	if latitude < -90 || latitude > 90 || longitude < -180 || longitude > 180 {
		return ErrInvalidCoordinates
	}

	// Проверяем существование пользователя
	existing, err := s.repo.GetByID(ctx, userID)
	if err != nil {
		s.logger.Error("Error checking user existence", zap.Error(err))
		return err
	}
	if existing == nil {
		return ErrUserNotFound
	}

	err = s.repo.UpdateLocation(ctx, userID, latitude, longitude)
	if err != nil {
		s.logger.Error("Error updating location", zap.Error(err))
		return err
	}

	s.logger.Info("User location updated successfully", zap.String("userID", userID))
	return nil
}

// GetMatches получает список потенциальных матчей
func (s *userService) GetMatches(ctx context.Context, userID string, latitude, longitude *float64, radiusKm float64, limit int) ([]model.User, error) {
	s.logger.Info("Getting matches for user",
		zap.String("userID", userID),
		zap.Float64p("latitude", latitude),
		zap.Float64p("longitude", longitude),
		zap.Float64("radiusKm", radiusKm),
		zap.Int("limit", limit),
	)

	// Проверяем существование пользователя
	existing, err := s.repo.GetByID(ctx, userID)
	if err != nil {
		s.logger.Error("Error checking user existence", zap.Error(err))
		return nil, err
	}
	if existing == nil {
		return nil, ErrUserNotFound
	}

	matches, err := s.repo.GetMatches(ctx, userID, latitude, longitude, radiusKm, limit)
	if err != nil {
		s.logger.Error("Error getting matches", zap.Error(err))
		return nil, err
	}

	s.logger.Info("Found matches", zap.String("userID", userID), zap.Int("count", len(matches)))
	return matches, nil
}
